core/aws-core/src/main/java/software/amazon/awssdk/awscore/endpoint/Sigv4aSigningRegionSetProvider.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.awscore.endpoint;
+
+
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.core.SdkSystemSetting;
+import software.amazon.awssdk.profiles.Profile;
+import software.amazon.awssdk.profiles.ProfileFile;
+import software.amazon.awssdk.profiles.ProfileFileSystemSetting;
+import software.amazon.awssdk.profiles.ProfileProperty;
+import software.amazon.awssdk.utils.Lazy;
+import software.amazon.awssdk.utils.StringUtils;
+import software.amazon.awssdk.utils.Validate;
+
+
+@SdkProtectedApi
+public final class Sigv4aSigningRegionSetProvider {
+    private final Supplier<ProfileFile> profileFile;
+    private final String profileName;
+    private final Lazy<Set<String>> regionSet;
+
+    private Sigv4aSigningRegionSetProvider(Builder builder) {
+        this.profileFile = Validate.paramNotNull(builder.profileFile, "profileFile");
+        this.profileName = builder.profileName;
+        this.regionSet = new Lazy<>(this::regionSet);
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public Set<String> resolveRegionSet() {
+        return this.regionSet.getValue();
+    }
+
+    /**
+     * Resolves the SIGV4A signing region set configuration.
+     * Returns a Set of non-empty strings if the configuration is set, or null if not  configured
+     */
+    private Set<String> regionSet() {
+        Optional<String> setting = SdkSystemSetting.AWS_SIGV4A_SIGNING_REGION_SET.getStringValue();
+        if (setting.isPresent()) {
+            return parseRegionSet(setting.get()).orElse(null);
+        }
+
+        ProfileFile file = this.profileFile.get();
+        Optional<Profile> profile = file.profile(profileName());
+        return profile
+            .flatMap(p -> p.property(ProfileProperty.SIGV4A_SIGNING_REGION_SET))
+            .flatMap(this::parseRegionSet)
+            .orElse(null);
+    }
+
+    private Optional<Set<String>> parseRegionSet(String value) {
```

Comment:
We don't normally use Optional with collection or map. Should we just return empty set if the value is blank?

Reply:
Agree , Updated to return Null directly if we donot find valid non empty values

Reply:
The reason we don't use Optional with collection is that we can use empty to avoid null handling. Should we return empty set in that case?

Reply:
Empty Set is like supporting "  " regionset
If we allow empty Set it like the set had a entry of empty Value.
The SEP says its a non empty value if it exist.
WDYT ?

Reply:
I meant using empty set to indicate no value is available so that we don't need to handle null. We can check whether it's empty in the caller code.

Reply:
Updated to return Empty Set also updated test case